### PR TITLE
ALTAPPS-1233: Shared, Android amplitude integration

### PR DESCRIPTION
**Pull request by XanderZhu**
**YouTrack Issues**:
[ALTAPPS-1233](https://vyahhi.myjetbrains.com/issue/ALTAPPS-1233)

**Checklist**

_Before Code Review:_

- [x] Fields "Assignees, Labels, Milestone" are filled in the pull request;
- [x] New analytics events are documented;
- [x] All checks have been passed;
- [x] Changes have been checked locally.

**Description**
- Create `AnalyticEventUserProperties` and pass it to `AnalyticEngine.reportEvent` from `AnalyticInteractor`; 
- Remove `setScreenOrientation` & `setAppTrackingTransparencyAuthorizationStatus` from  `AnalyticEngine` and leave them in `AnalyticInteractor`;
- Create `AmplitudeAnalyticEngine`;
- Implement `AndroidAmplitudeAnalyticEngine` & integrate Amplitude SDK;
- Make `AnalyticEvent.source` a `Set`; add `AMPLITUDE` to `HyperskillAnalyticEvent.source` to log it via `AmplitudeAnalyticEngine`.
- Add `AmplitudeAnalyticEventMapper` to map `HyperskillAnalyticEvent` to `AmplitudeAnalyticEvent`;
- Add new `AmplitudeAnalyticEvent`s;

### CODE REVIEW

File: shared/src/commonMain/kotlin/org/hyperskill/app/analytic/domain/interactor/AnalyticInteractor.kt [hunk 1]
```diff
@@ -24,17 +40,37 @@ class AnalyticInteractor(
...
     }
+
+    private suspend fun getUserProperties(): AnalyticEventUserProperties =
+        coroutineScope {
+            val profileDeferred = async {
+                currentProfileStateRepository.getState(forceUpdate = false).getOrNull()
+            }
+            val subscriptionDeferred = async {
+                currentSubscriptionStateRepository.getState(forceUpdate = false).getOrNull()
+            }
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/analytic/domain/interactor/AnalyticInteractor.kt:64]: I hope that `SharedPreferences` and `NSUserDefaults` has an optimization mechanism in place that involves caching, to not always perform a disk read operation whenever we request a value.

File: shared/src/commonMain/kotlin/org/hyperskill/app/analytic/domain/processor/AmplitudeAnalyticEventMapper.kt [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        when (analyticEvent) {
+            is HyperskillAnalyticEvent -> AmplitudeAnalyticEvent(
+                name = getType(analyticEvent),
+                params = getParams(analyticEvent)
+            )
+            is AmplitudeAnalyticEvent -> analyticEvent
+            else -> null
+        }
+
+    private fun getType(analyticEvent: HyperskillAnalyticEvent): String =
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/analytic/domain/processor/AmplitudeAnalyticEventMapper.kt:20]: I suppose result string should be with no spaces or new lines in beginning and end

**Review by ivan-magda — APPROVED** [review]

File: shared/src/androidMain/kotlin/org/hyperskill/app/analytic/domain/amplitude/AndroidAmplitudeAnalyticEngine.kt [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        force: Boolean
+    ) {
+        val amplitudeAnalyticEvent = AmplitudeAnalyticEventMapper.map(event)
+        if (amplitudeAnalyticEvent != null) {
+            amplitude.track(
+                BaseEvent().apply {
+                    eventType = amplitudeAnalyticEvent.name
+                    eventProperties = amplitudeAnalyticEvent.params.toMutableMap()
+                    this.userProperties = userProperties.asMapWithoutUserId().toMutableMap()
+                    this.userId = userProperties.userId?.toString()
```
[the remark this answers is not in the document]
Comment (XanderZhu) [reply shared/src/androidMain/kotlin/org/hyperskill/app/analytic/domain/amplitude/AndroidAmplitudeAnalyticEngine.kt:45]: @ivan-magda 
Setting userId via dedicated setUserId method makes the same things, that BaseEvent.userId. It is just a useful method to set userId once. As the userId for `AmplitudeAnalyticEngine` is taken from `CurrentProfileStateRepository` for every event, I decided to use the same solution for `AmlitudeAnalyticEngine`.

File: shared/src/commonMain/kotlin/org/hyperskill/app/analytic/domain/processor/AmplitudeAnalyticEventMapper.kt (hunk 2 above)
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/analytic/domain/processor/AmplitudeAnalyticEventMapper.kt:20]: @ivan-magda 
As for task description, "action, part & target should be concatenated using space char".
But, yeah, in some cases there will be a redundant space.

File: shared/src/commonMain/kotlin/org/hyperskill/app/analytic/domain/interactor/AnalyticInteractor.kt (hunk 1 above)
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/analytic/domain/interactor/AnalyticInteractor.kt:64]: @ivan-magda 
It's easy to add an in-memory cache to the appropriate state holders.

**Comment by XanderZhu** [comment]
@ivan-magda 
Please, take a look at the CurrentProfileStateHolderImpl & CurrentSubscriptionStateHolderImpl changes. I added an in-memory cache to reduce disc operations as we have many analytical events, and all of them require this data.

File: shared/src/commonMain/kotlin/org/hyperskill/app/analytic/domain/processor/AmplitudeAnalyticEventMapper.kt (hunk 2 above)
Comment (ivan-magda) [reply shared/src/commonMain/kotlin/org/hyperskill/app/analytic/domain/processor/AmplitudeAnalyticEventMapper.kt:20]: I suppose we should build string smth like this:
```kotlin
buildString {
    append(analyticEvent.action.actionName)
    if (analyticEvent.part != null) {
        append(' ')
        append(analyticEvent.part.partName)
    }
    if (analyticEvent.target != null) {
        append(' ')
        append(analyticEvent.target.targetName)
    }
}
```

Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/analytic/domain/processor/AmplitudeAnalyticEventMapper.kt:20]: @ivan-magda 
Yeah, your are right

**Pull request merged**